File: saltedge-sca-service-spring/saltedge-sca-service-sdk/src/test/java/com/saltedge/sca/sdk/controllers/ConnectionsControllerIntegrationTests.java
/*
 * This file is part of the Salt Edge Authenticator distribution
 * (https://github.com/saltedge/sca-identity-service-example).
 * Copyright (c) 2020 Salt Edge Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 or later.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * For the additional permissions granted for Salt Edge Authenticator
 * under Section 7 of the GNU General Public License see THIRD_PARTY_NOTICES.md
 */
package com.saltedge.sca.sdk.controllers;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.saltedge.sca.sdk.MockMvcTestAbs;
import com.saltedge.sca.sdk.TestTools;
import com.saltedge.sca.sdk.models.api.requests.ScaCreateConnectionRequest;
import com.saltedge.sca.sdk.models.api.responces.ScaCreateConnectionResponse;
import com.saltedge.sca.sdk.tools.DateTools;
import org.hamcrest.Matchers;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;

import static com.saltedge.sca.sdk.ScaSdkConstants.*;
import static com.saltedge.sca.sdk.controllers.ConnectionsController.CONNECTIONS_REQUEST_PATH;
import static com.saltedge.sca.sdk.tools.UrlTools.DEFAULT_AUTHENTICATOR_RETURN_TO;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.BDDMockito.given;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@RunWith(SpringRunner.class)
@WebMvcTest(ConnectionsController.class)
public class ConnectionsControllerIntegrationTests extends MockMvcTestAbs {
	@Test
	public void givenValidRequest_whenMakeCreateConnectionRequest_thenReturnOKAndRedirectToEnrollPage() throws Exception {
		//given
		given(connectionsService.createConnection(any(ScaCreateConnectionRequest.Data.class), isNull()))
				.willReturn(ScaCreateConnectionResponse.createResponseWithAuthorizeUrl("2", "https://localhost/admin/enroll?secret=auth_token"));

		ScaCreateConnectionRequest requestData = new ScaCreateConnectionRequest(new ScaCreateConnectionRequest.Data(
				publicKey,
				DEFAULT_AUTHENTICATOR_RETURN_TO,
				"android",
				"token",
				null
		));
		String json = new ObjectMapper().writeValueAsString(requestData);

		//when
		mvc.perform(post(CONNECTIONS_REQUEST_PATH)
				.contentType(MediaType.APPLICATION_JSON).accept(MediaType.APPLICATION_JSON)
				.content(json))

				//then
				.andExpect(status().isOk())
				.andExpect(jsonPath("$.data.id", Matchers.is("2")))
				.andExpect(jsonPath("$.data.connect_url", Matchers.is("https://localhost/admin/enroll?secret=auth_token")))
		;
	}

	@Test
	public void givenValidRequestWithConnectQueryParam_whenMakeCreateConnectionRequest_thenReturnOKAndReturnTo() throws Exception {
		//given
		given(connectionsService.createConnection(any(ScaCreateConnectionRequest.Data.class), eq("connectQuery")))
				.willReturn(ScaCreateConnectionResponse.createResponseWithAccessToken("1", "access_token"));
		ScaCreateConnectionRequest requestData = new ScaCreateConnectionRequest(new ScaCreateConnectionRequest.Data(
				publicKey,
				"authenticator//connect",
				"android",
				"token",
				"connectQuery"
		));
		String json = new ObjectMapper().writeValueAsString(requestData);

		//when
		mvc.perform(post(CONNECTIONS_REQUEST_PATH)
			.contentType(MediaType.APPLICATION_JSON).accept(MediaType.APPLICATION_JSON)
			.content(json))
			//then
			.andExpect(status().isOk())
			.andExpect(jsonPath("$.data.connect_url").doesNotExist())
			.andExpect(jsonPath("$.data.access_token", Matchers.is("access_token")))
			.andExpect(jsonPath("$.data.id", Matchers.is("1")));
	}

	@Test
	public void givenInvalidRequest_whenMakeCreateConnectionRequest_thenReturnBadRequest() throws Exception {
		//given
		ScaCreateConnectionRequest requestData = new ScaCreateConnectionRequest();

		//when
		mvc.perform(post(CONNECTIONS_REQUEST_PATH)
				.contentType(MediaType.APPLICATION_JSON).accept(MediaType.APPLICATION_JSON)
				.content(new ObjectMapper().writeValueAsString(requestData)))

		//then
				.andExpect(status().isBadRequest());
	}

	@Test
	public void revokeConnectionTest_returnSuccess() throws Exception {
		//given
		String requestUrl = "http://localhost" + CONNECTIONS_REQUEST_PATH;
		String expiresAt = String.valueOf((DateTools.nowUtcSeconds() + 60));
		String signature = TestTools.createSignature(
				HttpMethod.DELETE.toString(),
				requestUrl,
				expiresAt,
				"",
				TestTools.getRsaPrivateKey()
		);
		given(connectionsRepository.findByAccessTokenAndRevokedFalse("access_token")).willReturn(testAuthorizedConnection);

		//when
		mvc.perform(delete(CONNECTIONS_REQUEST_PATH)
				.accept(MediaType.APPLICATION_JSON)
				.header(HEADER_KEY_ACCESS_TOKEN, "access_token")
				.header(HEADER_KEY_EXPIRES_AT, expiresAt)
				.header(HEADER_KEY_SIGNATURE, signature)
		)
		//then
				.andExpect(status().isOk())
				.andExpect(jsonPath("$.data.access_token", Matchers.is("access_token")))
				.andExpect(jsonPath("$.data.success", Matchers.is(true)));
	}
}

File: spring-example/src/test/kotlin/com/saltedge/authenticator/identity/controller/api/ConnectionsControllerIntegrationTests.kt
/*
 * This file is part of the Salt Edge Authenticator distribution
 * (https://github.com/saltedge/sca-identity-service-example).
 * Copyright (c) 2019 Salt Edge Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 or later.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * For the additional permissions granted for Salt Edge Authenticator
 * under Section 7 of the GNU General Public License see THIRD_PARTY_NOTICES.md
 */
package com.saltedge.authenticator.identity.controller.api

import com.fasterxml.jackson.databind.ObjectMapper
import com.saltedge.authenticator.identity.HEADER_KEY_ACCESS_TOKEN
import com.saltedge.authenticator.identity.HEADER_KEY_EXPIRES_AT
import com.saltedge.authenticator.identity.HEADER_KEY_SIGNATURE
import com.saltedge.authenticator.identity.TestTools
import com.saltedge.authenticator.identity.model.Connection
import com.saltedge.authenticator.identity.model.ConnectionsRepository
import com.saltedge.authenticator.identity.model.User
import com.saltedge.authenticator.identity.model.mapping.CreateConnectionRequest
import com.saltedge.authenticator.identity.model.mapping.CreateConnectionRequestData
import com.saltedge.authenticator.identity.tools.nowUtcSeconds
import org.assertj.core.api.Assertions.assertThat
import org.hamcrest.Matchers
import org.hamcrest.Matchers.`is`
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.BDDMockito.given
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.HttpMethod
import org.springframework.http.MediaType
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status

@RunWith(SpringRunner::class)
@WebMvcTest(ConnectionsController::class)
class ConnectionsControllerIntegrationTests {
	@Autowired
	private val mvc: MockMvc? = null
	@MockBean
	private val repository: ConnectionsRepository? = null
	private val testConnection = Connection().apply {
		publicKey = TestTools.rsaPublicKeyString
		accessToken = "accessToken"
		user = User()
	}

	@Test
	fun createConnectionTest_returnSuccess() {
		val requestData = CreateConnectionRequest(CreateConnectionRequestData(
				publicKey = "key",
				returnUrl = "authenticator//connect",
				platform = "android",
				pushToken = "token"
		))
		val json: String = ObjectMapper().writeValueAsString(requestData)
		val connectionCaptor = ArgumentCaptor.forClass(Connection::class.java)
		mvc!!.perform(post(CONNECTIONS_REQUEST_PATH)
				.contentType(MediaType.APPLICATION_JSON).accept(MediaType.APPLICATION_JSON)
				.content(json))

				.andExpect(status().isOk)
				.andExpect(jsonPath("$.data.connect_url", Matchers.startsWith("https://localhost/admin/enroll?token=")))
				.andExpect(jsonPath("$.data.id", `is`("0")))

		Mockito.verify(repository!!).save(connectionCaptor.capture())
		val connection = connectionCaptor.value
		assertThat(connection.publicKey).isEqualTo(requestData.data!!.publicKey)
		assertThat(connection.pushToken).isEqualTo(requestData.data!!.pushToken)
		assertThat(connection.platform).isEqualTo(requestData.data!!.platform)
		assertThat(connection.returnUrl).isEqualTo(requestData.data!!.returnUrl)
		assertThat(connection.connectToken).isNotEmpty()
	}

	@Test
	fun createConnectionTest_returnError_whenInvalidRequestContent() {
		val requestData = CreateConnectionRequestData(
				publicKey = "key",
				returnUrl = "authenticator//connect",
				platform = "android",
				pushToken = "token"
		)

		mvc!!.perform(post(CONNECTIONS_REQUEST_PATH)
				.contentType(MediaType.APPLICATION_JSON).accept(MediaType.APPLICATION_JSON)
				.content(ObjectMapper().writeValueAsString(requestData)))

				.andExpect(status().isBadRequest)
	}

	@Test
	fun revokeConnectionTest_returnSuccess() {
		val requestUrl = "http://localhost$CONNECTIONS_REQUEST_PATH"
		val expiresAt = (nowUtcSeconds + 60).toString()
		val signature = TestTools.createSignature(
				requestMethod = "${HttpMethod.DELETE}",
				requestUrl = requestUrl,
				expiresAt = expiresAt,
				requestBody = "",
				privateKey = TestTools.rsaPrivateKey
		)

		given(repository!!.findByAccessTokenAndRevokedFalse("accessToken")).willReturn(testConnection)
		val connectionCaptor = ArgumentCaptor.forClass(Connection::class.java)

		mvc!!.perform(delete(CONNECTIONS_REQUEST_PATH)
				.accept(MediaType.APPLICATION_JSON)
				.header(HEADER_KEY_ACCESS_TOKEN, "accessToken")
				.header(HEADER_KEY_EXPIRES_AT, expiresAt)
				.header(HEADER_KEY_SIGNATURE, signature)
		)

				.andExpect(status().isOk)
				.andExpect(jsonPath("$.data.access_token", `is`("accessToken")))
				.andExpect(jsonPath("$.data.success", `is`(true)))

		Mockito.verify(repository).save(connectionCaptor.capture())
		assertThat(connectionCaptor.value.revoked).isTrue()
	}
}
